### PR TITLE
Migrate talentpool to new API schema

### LINKED ISSUE #1425: Allow multiple nominations on a member to be active at once
When a staff member runs the nominate command, if the user is already nominated it should not show an error message, instead it should accept the extra nomination. When a user is unnominated it should mark all current active nominations as inactive.

These changes should be reflected in a few places, `!tp list` and `!user` (in mod channels) should show number of current nominations.

### PR DESCRIPTION
**Pull request by ks129**
**Note: This has to be merged exactly same time than python-discord/site#447**

Migrates talent pool to the new system. More exact changes are in the commit message body.

Closes #1425

### CODE REVIEW

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 1]
```diff
@@ -101,8 +98,12 @@ async def watch_command(self, ctx: Context, user: FetchedMember, *, reason: str
         async with session.post(url, **kwargs) as resp:
             response_data = await resp.json()
 
-            if resp.status == 400 and response_data.get('user', False):
-                await ctx.send(":x: The specified user can't be found in the database tables")
+            if resp.status == 400:
+                if response_data.get('user', False):
+                    await ctx.send(":x: The specified user can't be found in the database tables")
+                elif response_data.get('actor', False):
+                    await ctx.send(":x: You already have nominated this user")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:105]: I think "You have already nominated this user" sounds better
```suggestion
                    await ctx.send(":x: You have already nominated this user")
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 2]
```diff
@@ -120,9 +121,7 @@ async def watch_command(self, ctx: Context, user: FetchedMember, *, reason: str
         )
 
         if history:
-            total = f"({len(history)} previous nominations in total)"
-            start_reason = f"Watched: {textwrap.shorten(history[0]['reason'], width=500, placeholder='...')}"
-            msg += f"\n\nUser's previous watch reasons {total}:```{start_reason}```"
+            msg += f"\n\n{len(history)} previous nominations in total"
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:124]: This should be in parentheses
```suggestion
            msg += f"\n\n({len(history)} previous nominations in total)"
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 3]
```diff
@@ -176,13 +175,39 @@ async def nomination_edit_group(self, ctx: Context) -> None:
 
     @nomination_edit_group.command(name='reason')
     @has_any_role(*MODERATION_ROLES)
-    async def edit_reason_command(self, ctx: Context, nomination_id: int, *, reason: str) -> None:
-        """
-        Edits the reason/unnominate reason for the nomination with the given `id` depending on the status.
+    async def edit_reason_command(self, ctx: Context, nomination_id: int, actor: FetchedMember, *, reason: str) -> None:
+        """Edits the reason of `actor` entry for the nomination with the given `id`."""
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:179]: The wording here is a little confusing, I'd go for something like:
```suggestion
        """Edits the reason of a specific nominator in a specific active nomination."""
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 4]
```diff
@@ -176,13 +175,39 @@ async def nomination_edit_group(self, ctx: Context) -> None:
...
+        try:
+            nomination = await self.bot.api_client.get(f"{self.api_endpoint}/{nomination_id}")
+        except ResponseCodeError as e:
+            if e.response.status == 404:
+                self.log.trace(f"Nomination API 404: Can't nomination with id {nomination_id}")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:184]: Should be
```suggestion
                self.log.trace(f"Nomination API 404: Can't edit a nomination with id {nomination_id}")
```

This seems to be a typo carried over from the original function. It seems to be in the function below as well, but I can't mark it since it's not a change.

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 5]
```diff
@@ -176,13 +175,39 @@ async def nomination_edit_group(self, ctx: Context) -> None:
...
+                await ctx.send(f":x: Can't find a nomination with id `{nomination_id}`")
+                return
+            else:
+                raise
 
-        If the nomination is active, the reason for nominating the user will be edited;
-        If the nomination is no longer active, the reason for ending the nomination will be edited instead.
-        """
+        if not nomination["active"]:
+            await ctx.send(":x: Can't edit reason of ended nomination.")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:191]: I suggest
```suggestion
            await ctx.send(":x: Can't edit the reason of an inactive nomination.")
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 6]
```diff
@@ -176,13 +175,39 @@ async def nomination_edit_group(self, ctx: Context) -> None:
...
+            return
+
+        if not any(entry["actor"] == actor.id for entry in nomination["entries"]):
+            await ctx.send(f":x: {actor} don't have entry for this nomination.")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:195]:
```suggestion
            await ctx.send(f":x: {actor} doesn't have an entry in this nomination.")
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 7]
```diff
@@ -176,13 +175,39 @@ async def nomination_edit_group(self, ctx: Context) -> None:
...
+            return
+
+        self.log.trace(f"Changing reason for nomination with id {nomination_id} of actor {actor} to {reason}")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:198]: I'd use the repr of the reason here
```suggestion
        self.log.trace(f"Changing reason for nomination with id {nomination_id} of actor {actor} to {repr(reason)}")
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 8]
```diff
@@ -176,13 +175,39 @@ async def nomination_edit_group(self, ctx: Context) -> None:
...
+
+        await self.bot.api_client.patch(
+            f"{self.api_endpoint}/{nomination_id}",
+            json={"actor": actor.id, "reason": reason}
+        )
+        await self.fetch_user_cache()  # Update cache
+        await ctx.send(":white_check_mark: Successfully updates reason of nomination.")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:205]:
```suggestion
        await ctx.send(":white_check_mark: Successfully updated nomination reason.")
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 9]
```diff
@@ -193,16 +218,18 @@ async def edit_reason_command(self, ctx: Context, nomination_id: int, *, reason:
             else:
                 raise
 
-        field = "reason" if nomination["active"] else "end_reason"
+        if nomination["active"]:
+            await ctx.send(":x: Cannot edit end reason of active nomination.")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:222]:
```suggestion
            await ctx.send(":x: Can't edit the end reason of an active nomination.")
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 10]
```diff
@@ -193,16 +218,18 @@ async def edit_reason_command(self, ctx: Context, nomination_id: int, *, reason:
...
+            return
 
-        self.log.trace(f"Changing {field} for nomination with id {nomination_id} to {reason}")
+        self.log.trace(f"Changing end reason for nomination with id {nomination_id} to {reason}")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:225]: A repr would be nice here as well
```suggestion
        self.log.trace(f"Changing end reason for nomination with id {nomination_id} to {repr(reason)}")
```

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 11]
```diff
@@ -237,13 +264,18 @@ async def unwatch(self, user_id: int, reason: str) -> bool:
...
         guild = self.bot.get_guild(Guild.id)
+        entries = []
+        for site_entry in nomination_object["entries"]:
+            actor_id = site_entry["actor"]
+            actor = guild.get_member(actor_id)
 
-        actor_id = nomination_object["actor"]
-        actor = guild.get_member(actor_id)
+            reason = site_entry["reason"] or "*None*"
+            created = time.format_infraction(site_entry["inserted_at"])
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:273]: This isn't really an infraction so maybe the name of the function should change, but this is out of scope for this PR

File: bot/exts/moderation/watchchannels/talentpool.py [hunk 12]
```diff
@@ -237,13 +264,18 @@ async def unwatch(self, user_id: int, reason: str) -> bool:
...
+            entries.append(f"Actor: {actor or actor_id}\nReason: {reason}\nCreated: {created}")
```
Comment (mbaruh) [inline bot/exts/moderation/watchchannels/talentpool.py:274]: Shouldn't this be `actor.mention`?

**Review by mbaruh — CHANGES_REQUESTED** [review]
Only a code review, haven't tested yet. Mostly grammer

File: bot/exts/moderation/watchchannels/talentpool.py (hunk 11 above)
Comment (ks129) [reply bot/exts/moderation/watchchannels/talentpool.py:273]: This would need to replace it everywhere, so yeah, let's not do this here.

**Review by wookie184 — APPROVED** [review]
Thanks for doing this! Have tested this and it seems to work as expected, nice job.

Only thing I would say is that the result from tp search is a bit messy/hard to read at the moment, here's an example
image
I'm not sure if changing this much is within the scope of this PR, if you think not I can add them in a separate PR, but here are some thoughts:
- The "Nomination ID" for inactive nominations should be after status and date like with active infractions.
- "Actor", "Created", "Reason" would be a better order than "Actor", "Reason", "Created".

**Comment by ks129** [comment]
@wookie184 I made these changes.

**Comment by mbaruh** [comment]
image
This doesn't paginate well, gets cut off. Doesn't seem to happen with the infractions paginator.

**Review by mbaruh — CHANGES_REQUESTED** [review]
After testing it seems to be working okay for many, a bit shorter nominations. So I guess limiting it to 1000 characters really is enough like you said.

I also think you should change the approval message in line 112 to ":white_check_mark: The nomination for {user} has been added to the talent pool"

**Review by mbaruh — APPROVED** [review]

**Review by mbaruh — APPROVED** [review]
